raster/src/main/scala/geotrellis/raster/render/BreakMap.scala

```diff
@@ -0,0 +1,133 @@
+package geotrellis.raster.render
+
+import geotrellis.raster._
+import geotrellis.util._
+
+import spire.algebra._
+import spire.math.Sorting
+import spire.std.any._
+import spire.syntax.order._
+
+import scala.specialized
+
+// --- //
+
+/** Root element in hierarchy for specifying the type of boundary when classifying colors*/
+sealed trait ClassBoundaryType
+case object GreaterThan extends ClassBoundaryType
+case object GreaterThanOrEqualTo extends ClassBoundaryType
+case object LessThan extends ClassBoundaryType
+case object LessThanOrEqualTo extends ClassBoundaryType
+case object Exact extends ClassBoundaryType
+
+/** A strategy for mapping values via a [[BreakMap]].
+  *
+  * '''Note:''' Specialized for `Int` and `Double`.
+  */
+class MapStrategy[@specialized(Int, Double) A](
+  val boundary: ClassBoundaryType,
+  val noDataValue: A,
+  val fallbackValue: A,
+  val strict: Boolean
+)
+
+/** Helper methods for constructing a [[MapStrategy]]. */
+object MapStrategy {
+  def int: MapStrategy[Int] = new MapStrategy(LessThanOrEqualTo, 0x00000000, 0x00000000, false)
+
+  def double: MapStrategy[Double] = new MapStrategy(LessThanOrEqualTo, Double.NaN, Double.NaN, false)
+}
+
+/** A `Map` which provides specific Binary Search-based ''map'' behaviour
+  * with breaks and a break strategy.
+  *
+  * {{{
+  * val bm: BreakMap[Int, Int] = ...
+  * val t: Tile = ...
+  *
+  * // Map all the cells of `t` to a target bin value in O(klogn).
+  * val newT: Tile = t.map(vm)
+  * }}}
+  *
+  * '''Note:''' `A` and `B` are specialized on `Int` and `Double`.
+  */
+class BreakMap[
+  @specialized(Int, Double) A: Order,
+  @specialized(Int, Double) B: Order
+](
+  breakMap: Map[A, B],
+  strategy: MapStrategy[B],
+  noDataCheck: A => Boolean
+) extends (A => B) {
+
+  /* A Binary Tree of the mappable values */
+  private lazy val vmTree: BTree[(A, B)] = {
+    val a: Array[(A, B)] = breakMap.toArray
+
+    Sorting.quickSort(a)
+
+    BTree.fromSortedSeq(a.toIndexedSeq).get
```

Comment:
`quickSort` needs an `Array`, but we could probably go faster if we avoided `toIndexedSeq` (makes it a `Vector`) and had another `BTree` constructor that took `Array`.

Within `fromSortedSeq`, we need the ability to `slice` and have O(1) access to elements.